sql: migrate planToString to new infrastructure

RaduBerinde opened the pull request:
#### sql: produce planComponents instead of planTop

Currently the `ConstructPlan` implementation produces a `planTop`, which is not
right conceptually; `planTop` encompasses higher-level plan state. There is a
hack that copies some of the existing state from the planner.

This change cleans this up by returning just the `*planComponents`.

Release note: None

#### sql: fold planInstrumentation back into planTop

The separation between `planTop` and `planInstrumentation` is not very useful or
clean (e.g. we have to pass `*planTop` down). This change folds the latter back
into the former.

Release note: None

#### sql: add fully and partially distributed plan flags

This change adds separate plan flags for the two distribution modes.

Release note: None

#### sql: migrate planToString to new infrastructure

This change reworks `planToString` (used to populate the plan file in statement
diagnostic bundles) to use the new explain infrastructure.

Release note: None

RaduBerinde committed:
sql: migrate planToString to new infrastructure

This change reworks `planToString` (used to populate the plan file in statement
diagnostic bundles) to use the new explain infrastructure.

Release note: None

RaduBerinde committed:
sql: produce planComponents instead of planTop

Currently the `ConstructPlan` implementation produces a `planTop`, which is not
right conceptually; `planTop` encompasses higher-level plan state. There is a
hack that copies some of the existing state from the planner.

This change cleans this up by returning just the `*planComponents`.

Release note: None

RaduBerinde committed:
sql: fold planInstrumentation back into planTop

The separation between `planTop` and `planInstrumentation` is not very useful or
clean (e.g. we have to pass `*planTop` down). This change folds the latter back
into the former.

Release note: None

RaduBerinde committed:
sql: add fully and partially distributed plan flags

This change adds separate plan flags for the two distribution modes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52883)
<!-- Reviewable:end -->

rohany commented on pkg/sql/distsql_running.go:
```diff
@@ -1056,7 +1056,7 @@ func (dsp *DistSQLPlanner) PlanAndRunCascadesAndChecks(
 			return false
 		}
-		cp := cascadePlan.(*planTop)
+		cp := cascadePlan.(*planComponents)
```
why are we casting these rather than just storing a planComponents?

RaduBerinde committed:
opt: fix bug in explain plan unwrapping

This change fixes a bug in the explain factory where we weren't unwrapping the
buffer node in exec.Cascade.

Release note: None

RaduBerinde commented:
___
*[pkg/sql/distsql_running.go, line 1059 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52883#-MEt00S1QYEsXnh8oysq-r4-1059:-MExwRcABGZH9WBxUVDz:b5ibzc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f4800d09676a97bdb28eeee2cdcebfb4db92fa9/pkg/sql/distsql_running.go#L1059)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

why are we casting these rather than just storing a planComponents?
</blockquote></details>

`PlanFn` returns `exec.Plan` which is an opaque interface (it's whatever the factory's `ConstructPlan` returns).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

bors r+
